Address Safer cpp failures in WebFrame

iidmsa opened the pull request:
#### c9cdb1f390be4b3f78fa3eeb4d1cef306896e891
<pre>
Address Safer cpp failures in WebFrame
<a href="https://bugs.webkit.org/show_bug.cgi?id=290742">https://bugs.webkit.org/show_bug.cgi?id=290742</a>
<a href="https://rdar.apple.com/problem/148226339">rdar://problem/148226339</a>

Reviewed by NOBODY (OOPS!).

Fix clang static analyzer warnings in WebFrame.

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::didFinishLoadInAnotherProcess):
(WebKit::WebFrame::didReceivePolicyDecision):
(WebKit::WebFrame::startDownload):
(WebKit::WebFrame::convertMainResourceLoadToDownload):
(WebKit::WebFrame::size const):
(WebKit::WebFrame::isMainFrame const):
(WebKit::WebFrame::jsContextForServiceWorkerWorld):
(WebKit::WebFrame::contentBounds const):
(WebKit::WebFrame::visibleContentBounds const):
(WebKit::WebFrame::scrollOffset const):
(WebKit::WebFrame::hasVerticalScrollbar const):
(WebKit::WebFrame::createSelectionSnapshot const):
(WebKit::WebFrame::handleContextMenuEvent):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9cdb1f390be4b3f78fa3eeb4d1cef306896e891

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7662 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102930 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48347 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17737 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25896 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74489 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31668 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100821 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13432 "Found 2 new test failures: http/tests/site-isolation/frame-access-after-window-open.html http/tests/site-isolation/iframe-top-level-navigation.html (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88405 "Found 16 new API test failures: TestWebKitAPI.SiteIsolation.OpenProvisionalFailure, TestWebKitAPI.SiteIsolation.ProtocolProcessSeparation, TestWebKitAPI.SiteIsolation.ChildNavigatingToNewDomain, TestWebKitAPI.SiteIsolation.ChildNavigatingToSameDomain, TestWebKitAPI.SiteIsolation.IframeOpener, TestWebKitAPI.SiteIsolation.LoadingCallbacksAndPostMessage, TestWebKitAPI.SiteIsolation.CancelProvisionalLoad, TestWebKitAPI.SiteIsolation.NavigateOpenerToProvisionalNavigationFailure, TestWebKitAPI.SiteIsolation.FindStringSelectionSameOriginFrames, TestWebKitAPI.SiteIsolation.ProvisionalLoadFailureOnCrossSiteRedirect ... (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54845 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13209 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47790 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83199 "Found 20 new API test failures: TestWebKitAPI.SiteIsolation.OpenProvisionalFailure, TestWebKitAPI.SiteIsolation.NavigationWithIFrames, TestWebKitAPI.SiteIsolation.ProtocolProcessSeparation, TestWebKitAPI.SiteIsolation.ChildNavigatingToNewDomain, TestWebKitAPI.SiteIsolation.ChildNavigatingToSameDomain, TestWebKitAPI.SiteIsolation.IframeOpener, TestWebKitAPI.SiteIsolation.FindStringSelection, TestWebKitAPI.SiteIsolation.LoadingCallbacksAndPostMessage, TestWebKitAPI.SiteIsolation.CancelProvisionalLoad, TestWebKitAPI.SiteIsolation.NavigateOpenerToProvisionalNavigationFailure ... (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6402 "Found 4 new test failures: http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/frame-access-after-window-open.html http/tests/site-isolation/iframe-top-level-navigation.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105310 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24898 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18134 "Found 4 new test failures: http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/frame-access-after-window-open.html http/tests/site-isolation/iframe-top-level-navigation.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83512 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25270 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82948 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27550 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5231 "Found 4 new test failures: http/tests/site-isolation/draw-after-navigation.html http/tests/site-isolation/draw-with-size-after-same-origin-navigation.html http/tests/site-isolation/frame-access-after-window-open.html http/tests/site-isolation/iframe-top-level-navigation.html (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/18505 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24859 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30028 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24681 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27995 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26255 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -608,13 +608,16 @@ void WebFrame::startDownload(const WebCore::ResourceRequest& request, const Stri
+    if (localFrame) {
+        Ref networkConnection = WebProcess::singleton().ensureNetworkProcessConnection();
+        networkConnection->connection().send(Messages::NetworkConnectionToWebProcess::StartDownload(policyDownloadID, request, topOrigin, isAppBound, suggestedName, fromDownloadAttribute, localFrame->frameID(), localFrame->pageID()), 0);
+    }
```
I doubt `networkConnection` needs the smart pointer here, since `connection()` is trivial and inlined. `send()` isn't though, so you probably need to protect `connection()`:
```suggestion
    if (localFrame)
        WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->send(Messages::NetworkConnectionToWebProcess::StartDownload(policyDownloadID, request, topOrigin, isAppBound, suggestedName, fromDownloadAttribute, localFrame->frameID(), localFrame->pageID()), 0); 
```

iidmsa replied:
Done!

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -608,13 +608,16 @@ void WebFrame::startDownload(const WebCore::ResourceRequest& request, const Stri
     }
     auto* localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    auto topOrigin = localFrame && localFrame->document() ? std::optional { localFrame->document()->topOrigin().data() } : std::nullopt;
+    RefPtr protectedLocalFrame { localFrame };
```
Remove `protectedLocalFrame` and just use `RefPtr` for `localFrame` above.

iidmsa replied:
Done!

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -624,7 +627,8 @@ void WebFrame::convertMainResourceLoadToDownload(DocumentLoader* documentLoader,
     }
     auto* localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    auto topOrigin = localFrame && localFrame->document() ? std::optional { localFrame->document()->topOrigin().data() } : std::nullopt;
+    RefPtr protectedLocalFrame { localFrame };
```
Same thing here - remove `protectedLocalFrame` and just use `RefPtr` for `localFrame` above.

iidmsa replied:
Done!

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -639,7 +643,8 @@ void WebFrame::convertMainResourceLoadToDownload(DocumentLoader* documentLoader,
     isAppBound = m_isNavigatingToAppBoundDomain;
-    webProcess.ensureNetworkProcessConnection().connection().send(Messages::NetworkConnectionToWebProcess::ConvertMainResourceLoadToDownload(mainResourceLoadIdentifier, policyDownloadID, request, topOrigin, response, isAppBound), 0);
+    Ref networkConnection = webProcess.ensureNetworkProcessConnection();
+    networkConnection->connection().send(Messages::NetworkConnectionToWebProcess::ConvertMainResourceLoadToDownload(mainResourceLoadIdentifier, policyDownloadID, request, topOrigin, response, isAppBound), 0);
```
`networkConnection` doesn't need `Ref`. Use `protectedConnection()` instead.

iidmsa replied:
Done!

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -708,18 +713,20 @@ String WebFrame::contentsAsString() const
 {
     RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    if (!localFrame)
+    RefPtr protectedLocalFrame { localFrame };
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -592,7 +592,7 @@ void WebFrame::didReceivePolicyDecision(uint64_t listenerID, PolicyDecision&& po
 
     if (policyDecision.policyAction == PolicyAction::Use && policyDecision.sandboxExtensionHandle) {
-        if (auto* page = this->page()) {
+        if (RefPtr page = this->page()) {
```
`page` doesn't need `RefPtr` here, since all functions called on `page` are trivial and inlined.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -788,10 +796,11 @@ String WebFrame::innerText() const
         return String();
 
-    if (!localFrame->document()->documentElement())
+    RefPtr protectedLocalFrame { localFrame };
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -833,28 +842,31 @@ Ref<API::Array> WebFrame::childFrames()
 {
     RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    if (!localFrame)
+    RefPtr protectedLocalFrame { localFrame };
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -833,28 +842,31 @@ Ref<API::Array> WebFrame::childFrames()
 {
     RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    if (!localFrame)
+    RefPtr protectedLocalFrame { localFrame };
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -833,28 +842,31 @@ Ref<API::Array> WebFrame::childFrames()
 {
     RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    if (!localFrame)
+    RefPtr protectedLocalFrame { localFrame };
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -877,20 +889,21 @@ JSGlobalContextRef WebFrame::jsContextForWorld(DOMWrapperWorld& world)
 JSGlobalContextRef WebFrame::jsContextForWorld(InjectedBundleScriptWorld* world)
 {
-    return jsContextForWorld(world->coreWorld());
+    return jsContextForWorld(RefPtr { world }->coreWorld());
```
This isn't needed, `world` should be protected before it's passed into this function.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -877,20 +889,21 @@ JSGlobalContextRef WebFrame::jsContextForWorld(DOMWrapperWorld& world)
 JSGlobalContextRef WebFrame::jsContextForServiceWorkerWorld(DOMWrapperWorld& world)
 {
-    if (!m_coreFrame || !m_coreFrame->page())
+    RefPtr coreFrame = m_coreFrame.get();
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1074,7 +1091,7 @@ JSValueRef WebFrame::jsWrapperForWorld(InjectedBundleCSSStyleDeclarationHandle*
         return nullptr;
 
-    auto* globalObject = localFrame->script().globalObject(world->coreWorld());
+    auto* globalObject = localFrame->script().globalObject(RefPtr { world }->coreWorld());
```
This isn't needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1086,7 +1103,7 @@ JSValueRef WebFrame::jsWrapperForWorld(InjectedBundleNodeHandle* nodeHandle, Inj
         return nullptr;
 
-    auto* globalObject = localFrame->script().globalObject(world->coreWorld());
+    auto* globalObject = localFrame->script().globalObject(RefPtr { world }->coreWorld());
```
Not needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1098,7 +1115,7 @@ JSValueRef WebFrame::jsWrapperForWorld(InjectedBundleRangeHandle* rangeHandle, I
         return nullptr;
 
-    auto* globalObject = localFrame->script().globalObject(world->coreWorld());
+    auto* globalObject = localFrame->script().globalObject(RefPtr { world }->coreWorld());
```
Not needed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1179,15 +1196,16 @@ void WebFrame::updateRemoteFrameSize(WebCore::IntSize size)
 {
     RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get());
-    if (!localFrame)
+    RefPtr protectedLocalFrame { localFrame };
```
Not needed.

iidmsa committed:
Address Safer cpp failures in WebFrame
https://bugs.webkit.org/show_bug.cgi?id=290742
rdar://problem/148226339

Reviewed by NOBODY (OOPS!).

Fix clang static analyzer warnings in WebFrame.

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::didFinishLoadInAnotherProcess):
(WebKit::WebFrame::didReceivePolicyDecision):
(WebKit::WebFrame::startDownload):
(WebKit::WebFrame::convertMainResourceLoadToDownload):
(WebKit::WebFrame::size const):
(WebKit::WebFrame::isMainFrame const):
(WebKit::WebFrame::jsContextForServiceWorkerWorld):
(WebKit::WebFrame::contentBounds const):
(WebKit::WebFrame::visibleContentBounds const):
(WebKit::WebFrame::scrollOffset const):
(WebKit::WebFrame::hasVerticalScrollbar const):
(WebKit::WebFrame::createSelectionSnapshot const):
(WebKit::WebFrame::handleContextMenuEvent):

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -890,7 +892,7 @@ JSGlobalContextRef WebFrame::jsContextForServiceWorkerWorld(DOMWrapperWorld& wor
 JSGlobalContextRef WebFrame::jsContextForServiceWorkerWorld(InjectedBundleScriptWorld* world)
 {
-    return jsContextForServiceWorkerWorld(world->coreWorld());
+    return jsContextForServiceWorkerWorld(RefPtr { world }->coreWorld());
```
This also isn't needed
